Raise TypeError when `__in`-operator used with a Document

malthejorgensen opened the pull request:
Currently doing a query like `BlogPost.objects(authors__in=author).count()`, where `author` is a single MongoEngine-`Document`, raises an obscure error:

```
raise ValidationError(message, errors=errors, field_name=field_name)
mongoengine.errors.ValidationError: u'id' is not a valid ObjectId, it must
be a 12-byte input or a 24-character hex string
```

This pull requests adds a "sanity" check, and raises a TypeError explaining that you need to use an iterable with the `in`-operator.

The correct thing to do is: `BlogPost.objects(authors__in=[author]).count()`, but that's not obvious when you read the `ValidationError` shown above.

<!-- Reviewable:start -->

[<img src="https://reviewable.io/review_button.svg" height="40" alt="Review on Reviewable"/>](https://reviewable.io/reviews/mongoengine/mongoengine/1237)

<!-- Reviewable:end -->

wojcikstefan commented on mongoengine/queryset/transform.py:
```diff
@@ -98,6 +99,15 @@ def query(_doc_cls=None, **kwargs):
                         value = value['_id']
 
             elif op in ('in', 'nin', 'all', 'near') and not isinstance(value, dict):
+                if isinstance(value, BaseDocument):
```
If all of the above truly expect an iterable, then we should raise an error for *any* non-iterable, not just a `BaseDocument`.

malthejorgensen replied:
Well, the problem is actually that `BaseDocument` _is_ iterable.

malthejorgensen replied:
Both the checks mentioned in this [StackOverflow thread](http://stackoverflow.com/questions/19919314/how-to-check-isinstance-of-iterable-in-python) will say that BaseDocument is iterable. (`for k in user` will over each field of the document `user`)

wojcikstefan replied:
Ahh, I missed that, thanks for clearing that up! I think in such case we can change this condition to:

```
# Raise an error if the in/nin/all/near param is not iterable. We need a
# special check for BaseDocument, because - although it's iterable - using
# it as such in the context of this method is most definitely a mistake.
if not hasattr(value, '__iter__') or isinstance(value, BaseDocument):
   raise TypeError(...)
```

The comment is crucial, too, given that this is a pretty unintuitive `isinstance` check.What do you think @malthejorgensen ?

malthejorgensen replied:
I agree, I'll make the change

wojcikstefan commented on tests/queryset/queryset.py:
```diff
@@ -4899,5 +4899,41 @@ class Doc(Document):
+            BlogPost.objects(authors__in=author).count()
+
+        # Check correct usage
+        try:
```
This block is redundant - we already check it above.

wojcikstefan commented on tests/queryset/queryset.py:
```diff
@@ -4899,5 +4899,41 @@ class Doc(Document):
+            self.fail("Using __in-operator raised type error when it shouldn't have")
+
+        User.drop_collection()
+        BlogPost.drop_collection()
```
These are redundant - by convention we should just take care of dropping existing collections at the beginning of each test (which you already do).

malthejorgensen commented:
I'll take a look at it now :)

malthejorgensen commented:
* I've added a separate test, so that there's now one for non-iterable values and one for `Document` values.
* I've removed the redundant part of the test (default behaviour of `__in`), and the redundant database cleanup (at the end of the test)
* I've changed the check so that there's a check for iterables and one for `Document` values

wojcikstefan commented:
@malthejorgensen you can now push the rebased version of the branch. Ping me when you do and I'll merge it. Thanks again!

malthejorgensen commented:
Absolutely no problem – I'm just glad to help! I'm only sorry that the process is turning out to be a bit of hassle for you.

I'm now the getting the **flake8** error `./mongoengine/queryset/transform.py:30:1: C901 'query' is too complex (44)` from Travis.

I have a different branch ([feature/in-operator-catch-no-list-flake-C901]) which should pass, where I've moved the value preparation out into its own function, but if we trust flake8's analysis, it's really just hiding an overarching problem – namely that the `query()`-function is getting too complex. And the decision of how to break it into smaller pieces, is a hard one.

We could also change the `max-complexity=42`-setting in `setup.cfg`, but that's also not ideal.

[feature/in-operator-catch-no-list-flake-C901]: https://github.com/malthejorgensen/mongoengine/tree/feature/in-operator-catch-no-list-flake-C901

wojcikstefan commented:
@malthejorgensen I apologize for all the back and forth, but since I merged #1428 a) there are conflicts again, b) `setup.cfg` increased allowed complexity to 45, which should cover your tweak (if it doesn't just increase it for now - I'll deal with simplifying it soon).

Also, please use `_import_class('BaseDocument')` inline in the method (e.g. like `CachedReferenceField`) so that we don't have to introduce a circular import.

malthejorgensen commented:
No problem @wojcikstefan – I'm just sorry to be taking your time for a small PR like mine
I think it should be ready now

wojcikstefan commented:
Thank you very much @malthejorgensen !

wojcikstefan commented:
FYI I made you tests a bit more concise while retaining all of the logic we want to test: 8e884fd
